keps/sig-node/20200310-ensure-secret-pulled-images.md

```diff
@@ -0,0 +1,169 @@
+- [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+## Summary
+
+We will add support for ensuring images pulled with pod imagePullSecrets are
+always authenticated even if cached. We will add a new boolean field
+`ensureSecretPulledImages` to the pod spec. The default to false
+means that if a first pod results in an image pulled with imagePullSecrets a
+second pod would have to be using always pull to ensure rights to use the
+previously pulled image. When set to true always pull would not be required,
+instead kublet will check if the image was pulled with an image pull secret and
+if so would force a pull of the image to ensure the image pulled with the
+secret is not used by another pod unless that pod also has the proper auth.
+
+## Motivation
+
+There have been customer requests for improving upon kubernetes ability to
+secure images pulled with auth. on a node. Issue
+[#18787](https://github.com/kubernetes/kubernetes/issues/18787) has been around
+for a while.
+
+To secure images one currently needs to inject `AllwaysPullImages` into pod
+specs via an admission plugin. As @liggitt [notes](https://github.com/kubernetes/kubernetes/issues/18787#issuecomment-532280931)
+the `pull` does not re-pull already-pulled layers of the image, but simply
+resolves/verifies the image manifest has not changed in the registry (which
+incidentally requires authenticating to private registries, which enforces the
+image access). That means in the normal case (where the image has not changed
+since the last pull), the request size is O(kb). However, the `pull` does put
+the registry in the critical path of starting a container, since an unavailable
+registry will fail the pull image manifest check (with or without proper
+authentication.)
+
+
+### Goals
+
+Add a flag processed by `kubelet` for `ensureSecretPulledImages` (or something
+similarly named) that, if true, would force `kubelet` to attempt to pull every
+image that was pulled with image pulled secret based authentication, regardless
+of the container image's pull policy.
+
+Optimize to only force re-authentication for a pod when the secret used to pull
+the container image is not present.
+
+### Non-Goals
+
+Out of scope for this KEP is an image caching policy that would direct container
+runtimes through the CRI wrt. how they should treat the caching of images on a
+node. Such as store for public use but only if encrypted. Or Store for private
+use unencrypted...
+
+## Proposal
+
+When `ensureSecretPulledImages` is set, `kublet` will check keep a list of
```

Comment:
One thing that I'm trying to think through is what happens if the `Kubelet` restarts, and this list gets cleared? Or will the list be persisted somewhere, and if so where?

Reply:
I was thinking on restart you would just have to reauthenticate at a minimum the manifest for each pod being created.. and that should be a lowish performance hit, or at least lower then pull always. I didn't want to checkpoint this and then implement some long running lifecycle to remove them.. exp. given our recent direction for shying away from checkpointing. IMO we can / should revisit that when a subsequent kep that implements a full cache policy over CRI to the runtime container image storage implementation(s).